packages/cozy-realtime/src/index.js

```diff
@@ -99,228 +120,233 @@ const configTypes = {
 
 const validateConfig = validate(configTypes)
 
-async function connectWebSocket(
+export function connectWebSocket(
   config,
   onmessage,
   onclose,
   numRetries,
-  retryDelay
+  retryDelay,
+  isRetry
 ) {
   validateConfig(config)
-  return new Promise((resolve, reject) => {
-    const options = {
-      secure: config.url ? isSecureURL(config.url) : true,
-      ...config
-    }
+  const options = {
+    secure: config.url ? isSecureURL(config.url) : true,
+    ...config
+  }
 
-    const protocol = options.secure ? 'wss:' : 'ws:'
-    const domain = options.domain || getDomainFromUrl(options.url)
+  const protocol = options.secure ? 'wss:' : 'ws:'
+  const domain = options.domain || new URL(options.url).host
 
-    if (!domain) {
-      throw new Error('Unable to detect domain')
-    }
+  if (!domain) {
+    throw new Error('Unable to detect domain')
+  }
 
-    const socket = new WebSocket(
-      `${protocol}//${domain}/realtime/`,
-      'io.cozy.websocket'
-    )
+  const socket = new WebSocket(
+    `${protocol}//${domain}/realtime/`,
+    'io.cozy.websocket'
+  )
 
-    socket.onopen = () => {
-      try {
-        socket.send(
-          JSON.stringify({
-            method: 'AUTH',
-            payload: options.token
-          })
-        )
-      } catch (error) {
-        return reject(error)
-      }
+  socket.onopen = () => {
+    try {
+      socket.send(
+        JSON.stringify({
+          method: 'AUTH',
+          payload: options.token
+        })
+      )
+    } catch (error) {
+      throw error
+    }
+  }
 
-      const windowUnloadHandler = () => socket.close()
-      window.addEventListener('beforeunload', windowUnloadHandler)
+  const windowUnloadHandler = () => socket.close()
+  window.addEventListener('beforeunload', windowUnloadHandler)
 
-      socket.onmessage = onmessage
-      socket.onclose = event => {
-        window.removeEventListener('beforeunload', windowUnloadHandler)
-        if (typeof onclose === 'function')
-          onclose(event, numRetries, retryDelay)
-      }
-      socket.onerror = error =>
-        // eslint-disable-next-line no-console
-        console.error(`WebSocket error: ${error.message}`)
+  socket.onmessage = onmessage
+  socket.onclose = event => {
+    window.removeEventListener('beforeunload', windowUnloadHandler)
+    if (typeof onclose === 'function') onclose(event, numRetries, retryDelay)
+  }
+  socket.onerror = error => console.error(`WebSocket error: ${error.message}`)
 
-      resolve(socket)
+  if (isRetry && subscriptionsState.size) {
+    for (let listenerKey of subscriptionsState) {
+      const { doctype, docId } = getTypeAndIdFromListenerKey(listenerKey)
+      subscribeWhenReady(doctype, socket, docId)
```

Comment:
It's probably not very important, but if we are trying to subscribe to a doctype while retrying to connect, we will call `subscribeWhenReady` from both here and line 273.

Reply:
Actually if there is already a listener registered for a doctype it won't. If for the previous subscription, we called unsubscribe, the listener will be removed and the new subscribe will send a new message as expected.